File: engines/node/levenshtein.js
const readline = require('readline')

const word  = process.argv[2]
const delta = parseInt(process.argv[3])
const limit = parseInt(process.argv[4])
const wordLen = word.length

function distance(a, b) {
  a = a.toLowerCase()
  b = b.toLowerCase()
  const m = a.length, n = b.length
  let prev = Array.from({ length: n + 1 }, (_, i) => i)
  let curr = new Array(n + 1)
  for (let i = 1; i <= m; i++) {
    curr[0] = i
    for (let j = 1; j <= n; j++) {
      curr[j] = a[i - 1] === b[j - 1]
        ? prev[j - 1]
        : 1 + Math.min(prev[j], curr[j - 1], prev[j - 1])
    }
    ;[prev, curr] = [curr, prev]
  }
  return prev[n]
}

const start = Date.now()
const results = []
const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity })

rl.on('line', line => {
  if (limit > 0 && results.length >= limit) return
  const candidate = line.trim()
  if (!candidate) return
  if (Math.abs(candidate.length - wordLen) <= delta && distance(word, candidate) === delta)
    results.push(candidate)
})

rl.on('close', () => {
  process.stdout.write(JSON.stringify({ duration_ms: Date.now() - start, results }) + '\n')
})

File: engines/python/levenshtein.py
import sys
import json
import time


def distance(a, b):
    a, b = a.lower(), b.lower()
    m, n = len(a), len(b)
    prev = list(range(n + 1))
    curr = [0] * (n + 1)
    for i in range(1, m + 1):
        curr[0] = i
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                curr[j] = prev[j - 1]
            else:
                curr[j] = 1 + min(prev[j], curr[j - 1], prev[j - 1])
        prev, curr = curr, prev
    return prev[n]


def main():
    word = sys.argv[1]
    delta = int(sys.argv[2])
    limit = int(sys.argv[3])
    word_len = len(word)

    start = time.time()
    results = []
    for line in sys.stdin:
        if limit > 0 and len(results) == limit:
            break
        candidate = line.rstrip("\n")
        if not candidate:
            continue
        if abs(len(candidate) - word_len) <= delta and distance(word, candidate) == delta:
            results.append(candidate)

    duration_ms = int((time.time() - start) * 1000)
    print(json.dumps({"duration_ms": duration_ms, "results": results}))


if __name__ == "__main__":
    main()
